criu/pie/restorer.c

```diff
@@ -340,6 +359,70 @@ static inline int restore_pdeath_sig(struct thread_restore_args *ta)
 	return 0;
 }
 
+/*
+ * Map an iovec with a given file,
+ * using `sys_mmap` instead of `sys_preadv`
+ * can accelerate restore process when
+ * there are some pages we won't touch later.
+ */
+static int map_iovec(struct iovec *iovs, loff_t off, int nr, struct task_restore_args *args)
+{
+	/* Using mmap to load content as an on-demand way rather than egarly read all the content */
+	int j;
+	VmaEntry *tmp_entry;
+	unsigned long iov_size, tmp;
+	loff_t local_offset = off;
+	ssize_t local_r, r = 0;
+
+	for (j = 0; j < nr; j++) {
+		/* p = iov_base and grow up, end = base + len */
+		unsigned long iov_offset = 0;
+		unsigned long end = iovs[j].iov_len;
+		pr_debug("Change preadv to map of iov buffer %d, iov base %p\n", j, iovs[j].iov_base);
+		/* Considering the iov buffer may cross multiple regions */
+		while (iov_offset < end) {
+			/* Find the vma_entry */
+			tmp_entry = lookup(iovs[j].iov_base + iov_offset, args);
+			iov_size = tmp_entry->end < (unsigned long)iovs[j].iov_base + end ?
+					   tmp_entry->end :
+						 (unsigned long)iovs[j].iov_base + end;
+			/* The length that we exactly want to map */
+			iov_size -= (unsigned long)iovs[j].iov_base + iov_offset;
+
+			/*
+			 * Now check whether it can be mapped, notice that some memory areas such as `vsyscall`,
+			 * which is not VMA_AREA_REGULAR, and stack area, whose flag has MAP_GROWSDOWN still need
+			 * to use `sys_pread`.
+			 */
+			if (!vma_entry_is(tmp_entry, VMA_AREA_REGULAR) || (tmp_entry->flags & MAP_GROWSDOWN) ||
+			    (tmp_entry->flags & MAP_ANONYMOUS) || (tmp_entry->flags & MAP_HUGETLB) ||
+			    (tmp_entry->flags & MAP_SHARED) || (tmp_entry->prot & PROT_WRITE)) {
```

Comment:
Did you test your changes? Here I see that we don't use mmap for anon and shared mappings. What other mappings do we expect here? Did I miss something?

Reply:
Why do we need to skip writable mappings?

Reply:
> Did you test your changes?


Sure. I've added them into CI test in my commits.

> Here I see that we don't use mmap for anon and shared mappings.

Sure, anonymous mappings and shared mappings are commonly used for IPC, and any updates are carried through to the underlying file. Remapping them into the `page.img` is not a good idea, I think.

> What other mappings do we expect here?

However, the flag `MAP_PRIVATE` is still ok. It won't affect the underlying file.

> Why do we need to skip writable mappings?

I'm doing this just to pass the CI test with `on-demand` mode. I explained the reason above.

Reply:
>> What other mappings do we expect here?

> However, the flag MAP_PRIVATE is still ok. It won't affect the underlying file.

Sorry. I didn't notice that MAP_PRIVATE isn't here. So do you want to handle only file private mappings this way? Do you have any real workload that benefits from this? Have you tested it?

>> Why do we need to skip writable mappings?

> I'm doing this just to pass the CI test with on-demand mode. I explained the reason above.

This is a bad explanation. CI passes because the introduced code path isn't used in tests. Read-only  private file mappings usually don't have any pages in images.

Reply:
>  So do you want to handle only file private mappings this way?

I don't want it, but it seems like remapping these shared and anonymous regions with `page.img` is a bad idea. 

> Do you have any real workload that benefits from this? Have you tested it?

Sure. For example, if you import some packages, like `import requests`, python will load all contents into memory while using a little part of it. I use CRIU to optimize my project about serverless, so removing these unused parts by using `sys_mmap` benefits.

> CI passes because the introduced code path isn't used in tests.

Nope. I've introduced it into the CI test in commit [d920aee](https://github.com/Yangfisher1/criu/commit/d920aee2e75c865255a3ce5bc550bfc38199dcdb). At first, It failed at `maps06` on CentOS 7. I fixed the failure by still choosing to use `sys_pread` for these `PROT_WRITE` regions, but I don't think it's necessary to check these permission bits as discussed [above](https://github.com/checkpoint-restore/criu/pull/1711#issuecomment-1047377556) .

Reply:
> I don't want it, but it seems like remapping these shared and anonymous regions with page.img is a bad idea.

Could you explain why it is bad?

> Sure. For example, if you import some packages, like import requests, python will load all contents into memory

Could you be more detailed here? How does it load all contents into memory? Does it create a read-only file mapping?  If the answer is yes,  how many pages does criu dump?

> Nope. I've introduced it into the CI test in commit d920aee. 

It is unclear for me whether tests triggers the new code path...

Could you show some numbers how it speeds up your workload? In ideal case, I would like to have an example that I can play with on my host.

Reply:
> Could you explain why it is bad?

For anonymous memory regions, it failed at the CI test, I explained the reason below.

> Could you be more detailed here? How does it load all contents into memory?

I'm not very familiar with this part but I'll search for some information. 

> Does it create a read-only file mapping? If the answer is yes, how many pages does criu dump?

Sure, mostly they're shared libraries like `.local/lib/python3.5/site-packages/Pillow.libs/libXdmcp-e15573e7.so.6.0.0` and `.local/lib/python3.5/site-packages/Pillow.libs/libopenjp2-b3d7668a.so.2.3.1`. The number of pages depends on applications and the packages that you import.

> It is unclear for me whether tests triggers the new code path...

file [scripts/ci/run-ci-tests.sh](https://github.com/checkpoint-restore/criu/commit/d920aee2e75c865255a3ce5bc550bfc38199dcdb#diff-00f0e392c1d1032a08a4204b60f0f3406cf270ed83bb0b8975e856bc01d5ed85), line 206.
```
./test/zdtm.py run -a -p 2 --on-demand --keep-going $ZDTM_OPTS
```
The option `--on-demand` makes it use on-demand mode.

> Could you show some numbers how it speeds up your workload? In ideal case, I would like to have an example that I can play with on my host.

Sure, I put two of my test cases on pastebin. Just call them TCA(Test Case A) and TCB(Test Case B).
TCA: https://pastebin.com/ujZjmrE5
TCB: https://pastebin.com/GXyr24Xf

We use a lock file to control the behavior of the program before dumping and after restoring. In our test, for TCA, the runtime is 10ms vs 5ms. For TCB, the runtime is 80ms vs 50ms.